Throwing a v128 should not clobber other stack slots.

justinmichaud opened the pull request:
#### e0e9244ff3e75eec3323507e98f435e780bd0264
<pre>
Throwing a v128 should not clobber other stack slots.
<a href="https://bugs.webkit.org/show_bug.cgi?id=250312">https://bugs.webkit.org/show_bug.cgi?id=250312</a>

Reviewed by Yusuke Suzuki.

Throwing a v128 is not currently supported, so we only allocate jsvalue-sized
stack slots. This is bad, because we still MoveVector, and this can cause
us to clobber important things. We should not clobber important things.

* JSTests/wasm/stress/simd-exception-throwing-v128-clobbers-fp.js: Added.
(catch):
* JSTests/wasm/stress/simd-exception-throwing-v128-clobbers-fp.wasm: Added.
* JSTests/wasm/stress/simd-exception-throwing-v128-clobbers-fp.wat: Added.
* JSTests/wasm/stress/simd-tail-calls-throw.js: Added.
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.func.start.export.string_appeared_here.call.f1.func.f1.local.v128.try.catch_all.return_call.f2.func.f2.async test):
* Source/JavaScriptCore/wasm/WasmAirIRGenerator64.cpp:
(JSC::Wasm::AirIRGenerator64::addThrow):
(JSC::Wasm::AirIRGenerator64::emitTailCallPatchpoint):
* Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp:
(JSC::Wasm::B3IRGenerator::addThrow):
(JSC::Wasm::B3IRGenerator::createTailCallPatchpoint):

Canonical link: <a href="https://commits.webkit.org/258789@main">https://commits.webkit.org/258789@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/66c37c053d9d4996ebaec0f3f9791d890222b643

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/102900 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/12019 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/35923 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/112151 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/172368 "Built successfully and passed tests") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/13035 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/2925 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/95142 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/109810 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/108674 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/10013 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/93206 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/37649 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/91858 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/24738 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/79385 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/93095 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/5464 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/26152 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/89469 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/3158 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/5615 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/2603 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/29685 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/11627 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/45644 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/92396 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6046 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/7363 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/20688 "Passed tests") | 
| | | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp:
```diff
@@ -3561,8 +3561,14 @@ auto B3IRGenerator::addThrow(unsigned exceptionIndex, Vector<ExpressionType>& ar
+        if (args[i]->type().isVector())
+            arg = constant(Double, 0);
+        patch->append(arg, ValueRep::stackArgument(i * sizeof(EncodedJSValue)));
+    }
```
Do we need the same thing in Air IR generator / Wasm LLInt?

justinmichaud replied:
Llint never sees SIMD functions, and we do have this in air

Constellation reviewed (approved):
r=me with comment.

justinmichaud commented on Source/JavaScriptCore/wasm/WasmAirIRGenerator64.cpp:
```diff
@@ -1986,7 +1986,11 @@ auto AirIRGenerator64::addThrow(unsigned exceptionIndex, Vector<ExpressionType>&
-        patchArgs.append(ConstrainedTmp(args[i], B3::ValueRep::stackArgument(i * sizeof(EncodedJSValue))));
+        // Nonetheless, they may clobber important things if they aren't treated as doubles.
+        auto arg = args[i];
+        if (args[i].type().isV128())
```
Here

Commit:
Throwing a v128 should not clobber other stack slots.
https://bugs.webkit.org/show_bug.cgi?id=250312

Reviewed by Yusuke Suzuki.

Throwing a v128 is not currently supported, so we only allocate jsvalue-sized
stack slots. This is bad, because we still MoveVector, and this can cause
us to clobber important things. We should not clobber important things.

* JSTests/wasm/stress/simd-exception-throwing-v128-clobbers-fp.js: Added.
(catch):
* JSTests/wasm/stress/simd-exception-throwing-v128-clobbers-fp.wasm: Added.
* JSTests/wasm/stress/simd-exception-throwing-v128-clobbers-fp.wat: Added.
* JSTests/wasm/stress/simd-tail-calls-throw.js: Added.
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.func.start.export.string_appeared_here.call.f1.func.f1.local.v128.try.catch_all.return_call.f2.func.f2.async test):
* Source/JavaScriptCore/wasm/WasmAirIRGenerator64.cpp:
(JSC::Wasm::AirIRGenerator64::addThrow):
(JSC::Wasm::AirIRGenerator64::emitTailCallPatchpoint):
* Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp:
(JSC::Wasm::B3IRGenerator::addThrow):
(JSC::Wasm::B3IRGenerator::createTailCallPatchpoint):

Canonical link: https://commits.webkit.org/258789@main
